amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+
+    annotations = PySequence_Fast(annotations, "expected sequence");
+    Py_ssize_t len = PySequence_Size(annotations);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        PyObject* pyAnnotation = PySequence_Fast_GET_ITEM(annotations, i);
+        Py_INCREF(pyAnnotation);
+        if (PyUnicode_Check(pyAnnotation)) {
+            ION_STRING annotation;
+            ion_string_from_py(pyAnnotation, &annotation);
+            err = ion_writer_add_annotation(writer, &annotation);
+        }
+        else if (PyObject_TypeCheck(pyAnnotation, (PyTypeObject*)_py_symboltoken_constructor)){
+            err = ionc_write_symboltoken(writer, pyAnnotation, /*is_value=*/FALSE);
+        }
+        Py_DECREF(pyAnnotation);
+        if (err) break;
+    }
+    Py_XDECREF(annotations);
+fail:
+    Py_XDECREF(annotations);
+    cRETURN;
+}
+
+/*
+ *  Writes a list or a sexp
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      sequence: An ion python list or sexp
+ *      tuple_as_sexp: Decides if a tuple is treated as sexp
+ *
+ */
+static iERR ionc_write_sequence(hWRITER writer, PyObject* sequence, PyObject* tuple_as_sexp) {
+    iENTER;
+    PyObject* child_obj = NULL;
+    sequence = PySequence_Fast(sequence, "expected sequence");
+    Py_ssize_t len = PySequence_Size(sequence);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        child_obj = PySequence_Fast_GET_ITEM(sequence, i);
+        Py_INCREF(child_obj);
+
+        IONCHECK(Py_EnterRecursiveCall(" while writing an Ion sequence"));
+        err = ionc_write_value(writer, child_obj, tuple_as_sexp);
+        Py_LeaveRecursiveCall();
+        IONCHECK(err);
+
+        Py_DECREF(child_obj);
+        child_obj = NULL;
+    }
+fail:
+    Py_XDECREF(child_obj);
+    Py_DECREF(sequence);
+    cRETURN;
+}
+
+/*
+ *  Writes a struct
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      map: An ion python struct
+ *      tuple_as_sexp: Decides if a tuple is treated as sexp
+ *
+ */
+static iERR ionc_write_struct(hWRITER writer, PyObject* map, PyObject* tuple_as_sexp) {
+    iENTER;
+    PyObject * list = PyMapping_Items(map);
+    PyObject * seq = PySequence_Fast(list, "expected a sequence within the map.");
+    PyObject * key = NULL, *val = NULL, *child_obj = NULL;
+    Py_ssize_t len = PySequence_Size(seq);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        child_obj = PySequence_Fast_GET_ITEM(seq, i);
+        key = PyTuple_GetItem(child_obj, 0);
+        val = PyTuple_GetItem(child_obj, 1);
+        Py_INCREF(child_obj);
+        Py_INCREF(key);
+        Py_INCREF(val);
+
+        if (PyUnicode_Check(key)) {
+            ION_STRING field_name;
+            ion_string_from_py(key, &field_name);
+            IONCHECK(ion_writer_write_field_name(writer, &field_name));
+        }
+        else if (key == Py_None) {
+            // if field_name is None, write symbol $0 instead.
+            IONCHECK(_ion_writer_write_field_sid_helper(writer, 0));
+        }
```

Comment:
Above `Py_None` indicates `$0` here. The if...else branch is for 'field_name' with `$0` rather than `null` value. The `null` field_name fails before reaching here. 

More specifically, below are some example:

```.py
obj = "{null: 12}"                      # This will fail, doesn't reach above branch
simpleion.dumps(simpleion.loads(obj), binary=False))    # amazon.ion.exceptions.IonException: IERR_INVALID_FIELDNAME


obj = "{$0: null}"            # This will reach above branch.
simpleion.dumps(simpleion.loads(obj), binary=False))     # {$0:null}
simpleion.dumps(simpleion.loads(obj), binary=True))      # b'\xe0\x01\x00\xea\xd2\x80\x0f'
```

Reply:
Got it, thanks.